core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1234,8 +1234,9 @@ class ReplicaManager(val config: KafkaConfig,
                                fetchOffset: Long,
                                currentTimeMs: Long): Option[Int] = {
     partition.leaderReplicaIdOpt.flatMap { leaderReplicaId =>
-      // Don't look up preferred for follower fetches via normal replication
-      if (Request.isValidBrokerId(replicaId))
+      // Don't look up preferred for follower fetches via normal replication and
+      // don't look up preferred read replica while fetch from follower replica
+      if (Request.isValidBrokerId(replicaId) || !partition.isLeader)
```

Comment:
I wonder if we should add the following method to `Partition` and use it instead of using `leaderReplicaIdOpt`. 

```scala
def leaderIdIfLocal: Option[Int] = {
  leaderReplicaIdOpt.filter(_ == localBrokerId)
}
```

We don't use any lock on this path so solely relying on `leaderReplicaIdOpt` to take the decision seems slightly better than calling `partition.isLeader`. What do you think?